### DISCUSSION TITLE
[Enhancement]: Evaluate if we need to implement rate-limiting for checkout in core.

### DISCUSSION THREAD

**Discussion by naman03malhotra** [woocommerce/woocommerce#49613] [General]
### Describe the solution you'd like

More context: p1720176681477009-slack-C0E1AV8T0

I also found that good folks from the blocks team are running a rate-limiting experiment ([Experiment: Add Rate Limits to Store API ](https://github.com/woocommerce/woocommerce-blocks/pull/5962/files)) for blocks checkout, which is behind a flag right now, [documentation](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/hooks/filters.md#woocommerce_store_api_rate_limit_options).

Their solution is based on tracking user ID and IP address and only allowing Y requests in X seconds.

However, we'd need to evaluate:
- Should such a solution exist in the core or be implemented at the payment plugin/gateway level?
- Do we want to extend the same solution to classic checkout and put it behind a feature flag (or make it default)?

Also, it would need some critical thinking and collecting feedback from the payment plugins, we can start with WC Payments and see if that will benefit them because most plugins apply rate-limits independently.

### Describe alternatives you've considered

Do nothing and no rate limits are applied; we rely on payment plugins to do that.

**Comment by maxrice** [comment]
We support & maintain the Clover payment gateway integration and we'd love to see this in core. It's a frequent issue in support and while Clover (and lots of other payment providers) have implemented their own solutions (like captchas, etc), it would be great for Woo to have a core option that can be part of a defense-in-depth strategy.

**Comment by barryhughes** [comment]
It would be great to widen this up to others, so I'm going to convert to a discussion. We'd also love input from @woocommerce/rubik given Naman's earlier note about [your rate limiting experiment](https://github.com/woocommerce/woocommerce-blocks/pull/5962/files) (ie, where should this live/is that sufficient)?

**Comment by senadir** [comment]
CC @wavvves who initially worked on rate limiting. 

My opinion is to yes, let's add this to the old Checkout as well, and enable it by default.

There's still some work that needs to be done from our side, like adding proper rate limiting on Batch requests.

**Comment by wavvves** [comment]
Having worked on this experiment, I strongly feel this is a **make or break** feature, and roll-out should be handled with care.
I'm totally on board with shipping this as a core feature, extending it to batch requests (the missing step to wrap Store API) and also to the old Checkout.

A few notes to think about:

- We can't safely assume how stores are set-up (eg behind reverse proxies etc) so I would be against having it enabled by default
- This is a feature for advanced users, shipping a UI to enable it must be accompanied with warnings, and links to documentation. It is a feature that either requires knowledge or guts to be experimented 😅. 
- At the time I made my best to have it compliant with [this RFC](https://datatracker.ietf.org/doc/html/rfc7239), probably we need to revisit it, and take extra care testing and reviewing it. 
- Also the original header-based IP approach can be questioned again.

It would be nice to check the adoption for the existing experiment 🤔

**Reply by senadir** [reply]
For most platforms this is enabled by default so I don't see any wrong in having it on by default.

**Reply by wavvves** [reply]
If a store with high purchase traffic is behind a reverse proxy or load balancer, the default rate limiting default settings will use that service IP address. This would rate limit all calls as a whole, throttling purchases for every user.
We can't ship it by default checking passthrough IP headers and using those if present, as that would make it trivial for a bad actor to spoof the IP and circumvent rate limiting 😞

**Reply by senadir** [reply]
We can make it on by default for new stores only for now.

I feel like a load balancer can pass back the IP just fine, we do check for several IPs before we actually see the remote address?

**Reply by wavvves** [reply]
We do if basic proxy support is enabled, but this has to be an option to be set. In theory we could be checking for passthrough IP headers to detect proxies, and if not present resort to remote_addr. But that would mean that an attacker supplying generated IPs through those headers would defeat the rate limiter 😞

**Reply by IanMisnerThough** [reply]
The issue of carding attacks impacts the perception of WooCommerce to a much wider breadth of merchants. The vast majority of non-technical users do not have the knowledge to even respond to them. This leads to people raising questions with their merchant services providers, who are large reputable companies who _directly blame WooCommerce_ in this scenario. 

While some large merchants might be inconvenienced, almost every other WooCommerce user has a significant risk (with both security and financial impact) resolved. 

Which is all to say: 

> This is a feature for advanced users, 

It's actually a feature for everyone else. Advanced users should be provided the tools to ensure their high-traffic setups can be maintained. Any store at the scale of having a reverse proxy or load balancer will be able to disable the default rate limiting behavior either through a setting or filter, with information to help them understand what impact that has over their store. 

I'd recommend default-enabled for all new users with dev docs/filters on how to disable and/or modify that behavior for high traffic stores.

**Comment by wavvves** [comment]
This might be relevant for the discussion:  p1722875752614459-slack-CRA4UEQQ3
